lib/Transforms/InsertGpuAllocs.cpp

```diff
@@ -0,0 +1,363 @@
+    // Checks the access type of the OP
+    auto getAccessType = [&](mlir::Value memref) {
+      AccessType ret;
+      for (auto mem : aliases.resolve(memref)) {
+        for (auto user : mem.getUsers()) {
+          if (mlir::isa<mlir::func::ReturnOp>(user)) {
+            ret.hostRead = true;
+            ret.hostWrite = true;
+            continue;
+          }
+
+          if (auto copy = mlir::dyn_cast<mlir::memref::CopyOp>(user)) {
+            if (copy.source() == mem)
+              ret.hostRead = true;
+
+            if (copy.target() == mem)
+              ret.hostWrite = true;
+
+            continue;
+          }
+
+          if (auto memInterface =
+                  mlir::dyn_cast<mlir::MemoryEffectOpInterface>(user)) {
+            bool onDevice = user->getParentOfType<mlir::gpu::LaunchOp>();
+            if (memInterface.hasEffect<mlir::MemoryEffects::Read>())
+              (onDevice ? ret.deviceRead : ret.hostRead) = true;
+
+            if (memInterface.hasEffect<mlir::MemoryEffects::Write>())
+              (onDevice ? ret.deviceWrite : ret.hostWrite) = true;
+
+            continue;
+          }
+          if (mlir::isa<mlir::func::CallOp>(user)) {
+            bool onDevice = user->getParentOfType<mlir::gpu::LaunchOp>();
+            (onDevice ? ret.deviceRead : ret.hostRead) = true;
+            (onDevice ? ret.deviceWrite : ret.hostWrite) = true;
+            continue;
+          }
+        }
+      }
+      return ret;
+    };
+
+    for (auto &it : gpuBufferAllocs) {
+      auto alloc = mlir::cast<mlir::memref::AllocOp>(it.first);
+      it.second = getAccessType(alloc);
+    }
+
+    auto &block = funcBody.front();
+    for (auto &it : gpuBufferParams) {
+      auto param = block.getArgument(it.first);
+      it.second = getAccessType(param);
+
+      it.second.hostRead = true;
+      it.second.hostWrite = true;
+    }
+
+    // GetMemrefGlobal Op Case:
+    // This is the case where the inputs are globals contants and accessed using
+    // memref.get_global op. This code will add the IR for memeory allocation on
+    // the device with gpu.alloc and insert a memref.copy from host to device.
+    mlir::OpBuilder builder(func);
+    llvm::SmallVector<mlir::Value> dims;
+    llvm::SmallPtrSet<mlir::Operation *, 8> filter;
+    auto term = block.getTerminator();
+    assert(term);
+
+    for (auto &it : gpuGetMemrefGlobalParams) {
+      auto getGlobalOp = mlir::cast<mlir::memref::GetGlobalOp>(it.first);
+      it.second = getAccessType(getGlobalOp);
+      it.second.hostRead = true;
+      it.second.hostWrite = true;
+
+      auto loc = getGlobalOp.getLoc();
+      auto access = it.second;
+      builder.setInsertionPointAfter(getGlobalOp);
+      auto getGlobalOpResult = getGlobalOp.getResult();
+      auto memrefType = getGlobalOp.getType().cast<mlir::MemRefType>();
+      auto rank = static_cast<unsigned>(memrefType.getRank());
+      filter.clear();
+      dims.clear();
+      for (auto i : llvm::seq(0u, rank)) {
```

Comment:
Does this cover shape with unknown rank, unkown dim, and known rank/dim? Do you have test cases for those?

Reply:
it works with known rank and unknown dims and obviously with static shapes ...not other cases..I will add a test case for dynamic dim case